### ISSUE TITLE
Blueprint routes does not support instance methods

### ISSUE THREAD

**Issue by metatoaster** [huge-success/sanic#659]
Given a class that provide some async method, with a registration method of some kind that will add a route to expose through a sanic instance like so:

```python
from sanic import Sanic
from sanic import Blueprint
from sanic import response


class Manager(object):

    async def info(self, request):
        return response.json({'id': id(self)})

    def register(self, app):
        app.add_route(self.info, '/info')


app = Sanic()
manager = Manager()
manager.register(app)
```

The registration will be done fine and the app will be able to serve the registered route (i.e. `/info`) as expected.  However if blueprints were involved, i.e. if the `Manager` class were to provide a blueprint and add the route to it instead, then have the app register the entire blueprint:

```python
class Manager(object):

    def __init__(self):
        self.blueprint = Blueprint(str(id(self)))
        self.blueprint.add_route(self.info, '/info')

    async def info(self, request):
        return response.json({'id': id(self)})

    def register(self, app):
        app.blueprint(self.blueprint)
```

The following exception will happen

```
Traceback (most recent call last):
  File "bpfail.py", line 21, in <module>
    manager.register(app)
  File "bpfail.py", line 16, in register
    app.blueprint(self.blueprint)
  File "lib/python3.6/site-packages/sanic/app.py", line 233, in blueprint
    blueprint.register(self, options)
  File "lib/python3.6/site-packages/sanic/blueprints.py", line 40, in register
    future.handler.__blueprintname__ = self.name
AttributeError: 'method' object has no attribute '__blueprintname__'
```

As Python does not appear to support assignment of arbitrary attributes to instance methods (but are allowed for standard functions and lambdas), this either need to be worked around or documented as unsupported.

**Comment by 38elements** [comment]
Is there needs of putting instance method in blueprint?

**Comment by metatoaster** [comment]
The fix could be as simple as catching the `AttributeError: 'method' object has no attribute '__blueprintname__'` and raising it as unsupported, or as I said, put it into the documentation as to why this can't be done, or use a workaround.

Also, when I realized what the intent of the marker was (also what it is trying to do), exposing the instance methods directly as a route is probably not as clean anyway. The better solution is to have a setup function to be the closure for the actual callables that will be the route handlers, which can wrap around the actual raw methods that provide the logic; being encapsulated in a closure also mean that the registration can be done for any number of blueprints.

The goal of this issue is to more document a potential problem, which hopefully will have a resolution that minimise any potential negative impacts that will be caused by this.

**Comment by Zheaoli** [comment]
@r0fls 
I suggest close this issue

**Comment by mashaalmemon** [comment]
@38elements I noticed that the project made heavy use of classes. I think to support those folks who enjoy programing with classes, supporting blueprint routes that are instance/class methods would be ideal. I wouldn't consider the issue big, but from my point of view I'd prefer to use instance methods in my routes as it keeps my code organized.

@metatoaster 's assessment is correct, re instance/class methods not supporting arbitrary attributes, where a function can. I will be submitting a pull request where for blueprint.routes, if an attribute cannot be added to a function, it is added to the instance/class to which the function belongs. I've tested. It appears to work in my environment and is backwards compatible.

@Zheaoli 's points in #667 are that an instance method shouldn't be used for routes. I'm ok with that although my pull request will allow this. In my use case I'm more interested in class methods which help me (as @Zheaoli mentioned) helps me keep my code organized. 

Hopefully this small change can make it to the next version of Sanic (0.7.1). Pull request is forthcoming.
